src/iterator/backend.rs

```diff
@@ -0,0 +1,380 @@
+        if lock[signal as usize].is_some() {
+            return Ok(());
+        }
+
+        let waker = Arc::clone(&self.waker);
+        let action = move || {
+            waker.pending[signal as usize].store(true, Ordering::SeqCst);
+            waker.wake();
+        };
+        let id = unsafe { crate::register(signal, action) }?;
+        lock[signal as usize] = Some(id);
+        Ok(())
+    }
+
+    /// Get a reference to the read end of the self pipe
+    pub fn get_read(&self) -> &R {
+        &self.waker.read
+    }
+
+    /// Drains all data from the internal self-pipe. This method will never block.
+    fn flush(&self) {
+        const SIZE: usize = 1024;
+        let mut buff = [0u8; SIZE];
+
+        unsafe {
+            // Draining the data in the self pipe. We ignore all errors on purpose. This
+            // should not be something like closed file descriptor. It could EAGAIN, but
+            // that's OK in case we say MSG_DONTWAIT. If it's EINTR, then it's OK too,
+            // it'll only create a spurious wakeup.
+            while libc::recv(
+                self.waker.read.as_raw_fd(),
+                buff.as_mut_ptr() as *mut libc::c_void,
+                SIZE,
+                libc::MSG_DONTWAIT,
+            ) > 0
+            {}
+        }
+
+        if self.is_closed() {
+            // Wake any other sleeping ends
+            // (if none wait, it'll only leave garbage inside the pipe, but we'll close it soon
+            // anyway).
+            self.waker.wake();
+        }
+    }
+
+    /// Returns an iterator of already received signals.
+    ///
+    /// This returns an iterator over all the signal numbers of the signals received since last
+    /// time they were read (out of the set registered by this `Signals` instance). Note that
+    /// they are returned in arbitrary order and a signal number is returned only once even if
+    /// it was received multiple times.
+    ///
+    /// This method returns immediately (does not block) and may produce an empty iterator if
+    /// there are no signals ready.
+    pub fn pending(&self) -> Pending {
+        if !self.is_closed() {
+            self.flush();
+        }
+        self.waker.pending()
+    }
+
+    /// Checks the reading end of the self pipe for available signals.
+    ///
+    /// If there are no signals available or this instance was already closed it returns
+    /// [`Option::None`]. If there are some signals it returns a [`Pending`](struct.Pending)
+    /// instance wrapped inside a [`Option::Some`]. However, due to implementation details,
+    /// this still can produce an empty iterator.
+    ///
+    /// This method doesn't check the reading end itself but uses the passed in callback. This
+    /// method blocks if and only if the callback blocks trying to read some bytes.
+    pub fn wait<F>(&self, has_signals: &mut F) -> Result<Option<Pending>, Error>
+    where
+        F: FnMut(&mut &R) -> Result<bool, Error>,
```

Comment:
To continue the discussion about the `&mut &R` (github, for some reason, doesn't allow me to continue there :-( ), it really does look weird in the code. The semantics would be that one can change the reference ‒ make it point somewhere else. So, basically the signature says that the passed-in callback can manipulate what reference this method uses internally. That's certainly not the intention of the code. Apart from that, this will not work with arbitrary `Read` so being generic over it seems weird (there's kind of not documented part of the requirement that `&R: Read`). This worked previously, because the `UnixStream` was private in here.

If we are OK with requiring the above (which seems to be common for all the things usable in here anyway), I'd propose changing the signature to just `&R`. Then, the implementation shouldn't need to put the `&mut` there explicitly, only accept the parameter as a mutable binding and the `.` will do its magic. AFAIK this should totally work:

```rust
fn check(mut r: &UnixStream) {
    let mut buf = [0];
    r.read(&mut buf); // The . auto-refs and adds the `&mut` here automatically.
}
```

The alternative would be to really pass `&mut R`, which would be the most API-correct thing to do. But it would also mean putting it into a mutex and I could guess that someone would be unhappy of having a (blocking) mutex in an async code. So maybe the first thing is somewhat better in practice.

Reply:
I also prefer passing a `&mut R` to the `has_signal` callback. Using a Mutex in the blocking Signals struct may be OK but for the async structs it's kinda weird. So I prefer a version without a Mutex. But this won't be possible without moving the reading end out of the Arc.

As far as I can see the main reason why the reading end is inside an Arc is the ability to clone Signals instances. Therefore I would like to re-post my thoughts from https://github.com/vorner/signal-hook/issues/39#issuecomment-670971359 at this point:

> As far as I can see there aren't many use cases for cloning the Signals struct. Especially because results will be random when cloning Signals instances and querying them from multiple threads. The most important use case I see is closing the Forever iterator or returning from the wait method early.
> 
> I would propose to remove cloning from the Signals struct and add a way for getting some CloseHandle which offers the close() method. CloseHandle must be Send and may be Clone as well.
> 
> This way it would be possible to store the reading pipe half outside of an Arc which would at least make the mio Source implementation simpler. But other implementation details may profit by this as well. It would also reduce the complexity of the closing mechanism because there can be at most one thread reading from the pipe. Yes, I'm looking at you flush() who has to write something to the pipe if in closed state ;-)
> 
> But this would be a rather big breaking change in the API. And I'm not sure if there are other use cases I'm not aware of.

Reply:
But I fear this makes the PR even larger. So I will implement your suggestion as a first step. Maybe we can investigate this in the future.

Reply:
One use case for cloning was the closing. But also adding more signals later on (I use that in `spirit`).

Some kind of splitting in two halves ‒ the `Signals` itself and a manipulation handle to close/add signals would be an interesting thing, kind of in line of how channels work in Rust.

But yes, that would definitely be good to put into a separate PR.